src/contracts/test/SmartOrder.sol

```diff
@@ -0,0 +1,125 @@
+    using GPv2SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    bytes32 public constant APPDATA = keccak256("SmartOrder");
+
+    address public immutable owner;
+    bytes32 public immutable domainSeparator;
+    IERC20 public immutable sellToken;
+    IERC20 public immutable buyToken;
+    uint256 public immutable totalSellAmount;
+    uint256 public immutable totalFeeAmount;
+    uint32 public immutable validTo;
+
+    bool public funded;
+
+    constructor(
+        GPv2Settlement settlement,
+        IERC20 sellToken_,
+        IERC20 buyToken_,
+        uint32 validTo_,
+        uint256 totalSellAmount_,
+        uint256 totalFeeAmount_
+    ) {
+        owner = msg.sender;
+        domainSeparator = settlement.domainSeparator();
+        sellToken = sellToken_;
+        buyToken = buyToken_;
+        validTo = validTo_;
+        totalSellAmount = totalSellAmount_;
+        totalFeeAmount = totalFeeAmount_;
+
+        sellToken_.approve(
+            address(settlement.allowanceManager()),
+            totalSellAmount_.add(totalFeeAmount_)
+        );
+    }
+
+    modifier onlyOwner {
+        require(msg.sender == owner, "not owner");
+        _;
+    }
+
+    function fund() external onlyOwner {
+        require(!funded, "already funded");
+        funded = true;
+        sellToken.safeTransferFrom(
+            owner,
+            address(this),
+            totalSellAmount.add(totalFeeAmount)
+        );
+    }
+
+    function close() external onlyOwner {
+        uint256 balance = sellToken.balanceOf(address(this));
+        if (balance != 0) {
+            sellToken.safeTransfer(owner, balance);
+        }
+        selfdestruct(payable(owner));
+    }
+
+    function isValidSignature(bytes32 hash, bytes memory signature)
+        external
+        view
+        override
+        returns (bytes4 magicValue)
+    {
+        uint256 sellAmount = abi.decode(signature, (uint256));
+        GPv2Order.Data memory order = orderForSellAmount(sellAmount);
+
+        if (order.hash(domainSeparator) == hash) {
+            magicValue = GPv2EIP1271.MAGICVALUE;
+        }
+    }
+
+    function orderForSellAmount(uint256 sellAmount)
+        public
+        view
+        returns (GPv2Order.Data memory order)
+    {
+        uint256 balance = sellToken.balanceOf(address(this));
+        uint256 soldAmount =
+            totalSellAmount.sub(
+                balance.mul(totalSellAmount).div(
+                    totalSellAmount.add(totalFeeAmount)
+                )
+            );
+
+        order.sellToken = sellToken;
+        order.buyToken = buyToken;
+        order.receiver = owner;
+        order.sellAmount = sellAmount;
+        order.buyAmount = buyAmountForSellAmount(sellAmount, soldAmount);
+        order.validTo = validTo;
+        order.appData = APPDATA;
+        order.feeAmount = totalFeeAmount.mul(sellAmount).div(totalSellAmount);
+        order.kind = GPv2Order.SELL;
+        order.partiallyFillable = false;
```

Comment:
It can be partially fillable as well. It would just not make sense for the solver to chose any other `sellAmount` than will be completely filled as it would increase the limit price (and reduce price overlap and utility). Since partially fillable orders require a little more math (and gas), I thought it made more sense to produce fill-or-kill orders, as "smart" solver would fully fill the orders anyway.

Reply:
Added comment explaining why we set `partiallyFillable` to false.